File: shared/FitFusionCore/Sources/FitFusionCore/CrashReportingService.swift
import Foundation
#if canImport(Sentry)
import Sentry
#endif

/// Privacy-first wrapper around Sentry. The SDK is **only** initialized when
/// BOTH conditions are true:
///   * The user has opted in via Settings -> "Send crash reports"
///   * A build-time `SENTRY_DSN` is configured (Info.plist key or env var)
///
/// When either condition is false, every method here is a no-op. No data
/// ever leaves the device. Default state is **disabled** so Guest Mode users
/// don't get any background telemetry by surprise.
@MainActor
public final class CrashReportingService: ObservableObject {

    public static let shared = CrashReportingService()

    public static let optInKey = "crashReportsEnabled"

    /// Persisted opt-in flag. Surface this via a Settings toggle.
    @Published public var isEnabled: Bool {
        didSet { UserDefaults.standard.set(isEnabled, forKey: Self.optInKey) }
    }

    private(set) public var isStarted: Bool = false

    private init() {
        self.isEnabled = UserDefaults.standard.bool(forKey: Self.optInKey)
    }

    /// Read the DSN from (in order): the env var `SENTRY_DSN`, then the
    /// Info.plist key `SentryDSN`. Returns `nil` if neither is set.
    public static func resolveDSN() -> String? {
        if let env = ProcessInfo.processInfo.environment["SENTRY_DSN"],
           !env.isEmpty {
            return env
        }
        if let plistDSN = Bundle.main.object(forInfoDictionaryKey: "SentryDSN")
            as? String, !plistDSN.isEmpty {
            return plistDSN
        }
        return nil
    }

    /// Call from `FitFusionApp.init` (or `.task` after auth is restored).
    /// Safe to call multiple times \u2014 it short-circuits when already started.
    public func bootstrapIfEnabled(release: String,
                                   environment: String = "production") {
        #if canImport(Sentry)
        guard isEnabled, !isStarted, let dsn = Self.resolveDSN() else { return }
        SentrySDK.start { options in
            options.dsn = dsn
            options.releaseName = release
            options.environment = environment
            // Privacy: never collect PII; sample 100% errors but ZERO traces.
            options.tracesSampleRate = 0
            options.attachStacktrace = true
            options.sendDefaultPii = false
            options.maxBreadcrumbs = 50
            options.beforeSend = { event in
                // Strip any user-identifiable email-shaped strings
                event.user = nil
                return event
            }
        }
        isStarted = true
        #endif
    }

    /// Toggle from Settings. Stops the SDK if user opts out at runtime.
    public func setEnabled(_ enabled: Bool) {
        isEnabled = enabled
        #if canImport(Sentry)
        if !enabled, isStarted {
            SentrySDK.close()
            isStarted = false
        }
        #endif
    }

    /// Manual capture for non-fatal errors caught by the app.
    public func captureError(_ error: Error,
                             extras: [String: Any]? = nil) {
        #if canImport(Sentry)
        guard isStarted else { return }
        SentrySDK.capture(error: error) { scope in
            extras?.forEach { scope.setExtra(value: $0.value, key: $0.key) }
        }
        #endif
    }

    /// Manual capture for breadcrumb-style messages.
    public func captureMessage(_ message: String,
                               level: SentryLevelMirror = .info) {
        #if canImport(Sentry)
        guard isStarted else { return }
        let sentryLevel: SentryLevel
        switch level {
        case .debug:   sentryLevel = .debug
        case .info:    sentryLevel = .info
        case .warning: sentryLevel = .warning
        case .error:   sentryLevel = .error
        case .fatal:   sentryLevel = .fatal
        }
        SentrySDK.capture(message: message) { scope in
            scope.setLevel(sentryLevel)
        }
        #endif
    }
}

/// Sentry-independent level enum so callers don't have to `import Sentry`.
public enum SentryLevelMirror {
    case debug, info, warning, error, fatal
}

File: apps/HealthApp/android/app/src/main/java/com/myhealth/app/crash/CrashReportingService.kt
package com.myhealth.app.crash

import android.content.Context
import android.content.SharedPreferences
import io.sentry.Sentry
import io.sentry.SentryLevel
import io.sentry.android.core.SentryAndroid

/**
 * Privacy-first wrapper around the Sentry Android SDK. Mirrors the iOS
 * `CrashReportingService` contract:
 *
 *   * Disabled by default. Only initializes when the user opts in via
 *     Settings -> "Send crash reports".
 *   * Reads the DSN from `BuildConfig.SENTRY_DSN` (injected at build time)
 *     OR from the `SENTRY_DSN` env var. No-op if neither is set.
 *   * Strips PII before send. Sample rate for transactions is 0.
 */
object CrashReportingService {

    private const val PREFS = "myhealth_crash"
    private const val KEY_ENABLED = "enabled"
    @Volatile private var started: Boolean = false

    fun isEnabled(context: Context): Boolean =
        prefs(context).getBoolean(KEY_ENABLED, false)

    fun setEnabled(context: Context, enabled: Boolean) {
        prefs(context).edit().putBoolean(KEY_ENABLED, enabled).apply()
        if (!enabled && started) {
            Sentry.close()
            started = false
        }
    }

    /** Call from Application.onCreate. No-op when disabled or DSN missing. */
    fun bootstrapIfEnabled(context: Context, releaseName: String,
                           environment: String = "production") {
        if (!isEnabled(context) || started) return
        val dsn = resolveDsn() ?: return
        SentryAndroid.init(context) { options ->
            options.dsn = dsn
            options.release = releaseName
            options.environment = environment
            options.tracesSampleRate = 0.0
            options.isAttachStacktrace = true
            options.isSendDefaultPii = false
            options.maxBreadcrumbs = 50
            options.beforeSend = io.sentry.SentryOptions.BeforeSendCallback { event, _ ->
                event.user = null  // strip any user-identifiable data
                event
            }
        }
        started = true
    }

    fun captureError(t: Throwable) {
        if (started) Sentry.captureException(t)
    }

    fun captureMessage(msg: String, level: SentryLevel = SentryLevel.INFO) {
        if (started) Sentry.captureMessage(msg, level)
    }

    private fun prefs(context: Context): SharedPreferences =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    private fun resolveDsn(): String? {
        // BuildConfig.SENTRY_DSN is injected by Gradle when the
        // SENTRY_DSN env var or local property is set; fall back to the
        // process env so dev runs work without a Gradle rebuild.
        val fromEnv = System.getenv("SENTRY_DSN")
        if (!fromEnv.isNullOrEmpty()) return fromEnv
        return try {
            val cls = Class.forName("com.myhealth.app.BuildConfig")
            val field = cls.getField("SENTRY_DSN")
            (field.get(null) as? String).takeIf { !it.isNullOrEmpty() }
        } catch (_: Throwable) { null }
    }
}
